### PR TITLE
feat: Add support for OPA authorizer

### LINKED ISSUE #400: Implement Authorizer
In order to allow authorizing HDFS access requests with OPA, we need to implement a component to run inside of HDFS and forward requests to OPA, this is a principle very similar to the ones we currently use in [Kafka](https://github.com/StyraInc/opa-kafka-plugin), [Druid](https://github.com/stackabletech/druid-opa-authorizer) and which is being merged in upstream [Trino](https://github.com/trinodb/trino/pull/19532) as well. 

There is some code at https://github.com/stackabletech/hdfs-opa-authorizer where I have played around a bit, but none of this should be taken as gospel, if it seems weird then that is because it is weird and wasn't thought through!

The authorizer should implement HDFS' internal authorizer interface, serialize the relevant information from the context and forward it to a configurable OPA server.

Configuration could either be done via the HDFS config mechanism or in a first stage via environment variables.

```[tasklist]
### Tasks
- [x] Initial release of HDFS authorizer
- [x] Add HDFS authorizer to Stackable HDFS docker image
- [x] Update [authorization section](https://docs.stackable.tech/home/stable/hdfs/usage-guide/security#_authorization) in the docs. Document the JSON object that comes in when writing rules. ideally also give some example rules.
- [ ] https://github.com/stackabletech/docker-images/pull/555
- [x] feature tracker updated
```

### PR DESCRIPTION
**Pull request by sbernauer**
# Description

Closes https://github.com/stackabletech/hdfs-operator/issues/400

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [x] Changes are OpenShift compatible
- [x] CRD changes approved
- [x] CRD documentation for all fields, following the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs-style-guide).
- [x] Integration tests passed (for non trivial changes)
- [x] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [x] Code contains useful comments
- [x] Code contains useful logging statements
- [x] (Integration-)Test cases added
- [x] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs-style-guide).
- [x] Changelog updated
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

### CODE REVIEW

File: rust/crd/src/lib.rs
Comment (sbernauer) [inline rust/crd/src/lib.rs]: marker

File: rust/operator-binary/src/security/opa.rs
Comment (NickLarsenNZ) [inline rust/operator-binary/src/security/opa.rs:1]: suggestion (non-blocker): add doc-comments to the methods in the `HdfsOpaConfig` impl?
For example, I wasn't quite sure how the `add_hdfs_site_configs` and `add_core_site_configs` differ.

**Review by NickLarsenNZ — DISMISSED** [review]
LGTM

File: tests/templates/kuttl/kerberos/12-rego-rules.yaml
Comment (sbernauer) [inline tests/templates/kuttl/kerberos/12-rego-rules.yaml]: Marker to update before merging

**Comment by sbernauer** [comment]
In terms of CRD change we have two options:

### 1. Enable authorizer and group-mapper simultaneous

```yaml
  # Enable authorizer and group-mapper at the same time
  clusterConfig:
    authorization: # optional
      opa: # mandatory
        configMapName: opa # mandatory
        package: hdfs # mandatory
```

* Good, because consistent and users can not enable authZ and forget about group mapping
* Good, because rego rules can rely on the groups being propagated (although not recommended)

### 2. Enable authorizer and group-mapper separately

```yaml
  clusterConfig:
    authorization: # optional
      opaAuthorization: # mandatory
        configMapName: opa # mandatory
        package: hdfs # mandatory
      opaGroupMapping: # optional
        configMapName: opa # mandatory
        package: hdfs # mandatory
```

* Good, because more flexible, e.g. you can enable AuthZ without group mapping (which you basically would get for free)
* Bad, because more complex and error-prone

Originally I was thinking of 2., but now I am in favor of 1., as it's simpler and more consistent and 2. only enables stuff we should probably not support :)
User can always use configOverrides to easily partially enabled stuff when they really really want to.

**Comment by sbernauer** [comment]
Started https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/hdfs-operator-it-custom/128/

File: rust/operator-binary/src/security/opa.rs
Comment (sbernauer) [reply rust/operator-binary/src/security/opa.rs:1]: It probably won't help much, but I added a very basic sentence in 7af2ecc :see_no_evil:

File: docs/modules/hdfs/pages/usage-guide/security.adoc
Comment (adwk67) [inline docs/modules/hdfs/pages/usage-guide/security.adoc]:
```suggestion
This rego rule is intended for demonstration purposes and allows every operation.
```

File: docs/modules/hdfs/pages/usage-guide/security.adoc
Comment (adwk67) [inline docs/modules/hdfs/pages/usage-guide/security.adoc]:
```suggestion
For a production setup you probably want to take a look at our integration tests for a more secure set of rego rules.
```

File: docs/modules/hdfs/pages/usage-guide/security.adoc
Comment (adwk67) [inline docs/modules/hdfs/pages/usage-guide/security.adoc]:
```suggestion
In short, the current rego rules ignore the file ownership, permissions, ACLs and all other attributes files can have.
```

File: docs/modules/hdfs/pages/usage-guide/security.adoc
Comment (adwk67) [inline docs/modules/hdfs/pages/usage-guide/security.adoc]:
```suggestion
All of this is state in HDFS and clashes with the infrastructure-as-code approach (IaC).
```

File: docs/modules/hdfs/pages/usage-guide/security.adoc
Comment (adwk67) [inline docs/modules/hdfs/pages/usage-guide/security.adoc]:
```suggestion
Instead, HDFS will send a request detailing who (e.g. `alice/test-hdfs-permissions.default.svc.cluster.local@CLUSTER.LOCAL`) is trying to do what (e.g. `open`, `create`, `delete` or `append`) on what file (e.g. `/foo/bar`).
```

File: docs/modules/hdfs/pages/usage-guide/security.adoc
Comment (adwk67) [inline docs/modules/hdfs/pages/usage-guide/security.adoc]:
```suggestion
OPA then makes a decision if this action is allowed or not.
```

File: docs/modules/hdfs/pages/usage-guide/security.adoc
Comment (adwk67) [inline docs/modules/hdfs/pages/usage-guide/security.adoc]:
```suggestion
Instead of `chown`-ing a directory to a different user to assign write permissions, you should go to your IaC Git repository and add a rego rule entry specifying that the user is allowed to read and write to that directory.
```

File: docs/modules/hdfs/pages/usage-guide/security.adoc
Comment (adwk67) [inline docs/modules/hdfs/pages/usage-guide/security.adoc]:
```suggestion
We encountered several challenges while implementing the group mapper, the most serious of which being that the `GroupMappingServiceProvider` interface only passes the `shortUsername` when https://github.com/apache/hadoop/blob/a897e745f598ef05fc0c253b2a776100e48688d2/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/GroupMappingServiceProvider.java#L45[asking for group memberships].
```

File: docs/modules/hdfs/pages/usage-guide/security.adoc
Comment (adwk67) [inline docs/modules/hdfs/pages/usage-guide/security.adoc]:
```suggestion
This does not allow us to differentiate between e.g. `hbase/hbase-prod.prod-namespace.svc.cluster.local@CLUSTER.LOCAL` and `hbase/hbase-dev.dev-namespace.svc.cluster.local@CLUSTER.LOCAL`, as the GroupMapper will get asked for `hbase` group memberships in both cases.
```

File: docs/modules/hdfs/pages/usage-guide/security.adoc
Comment (adwk67) [inline docs/modules/hdfs/pages/usage-guide/security.adoc]:
```suggestion
Users could work around this to assign unique shortNames by using https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-common/SecureMode.html#Mapping_from_Kerberos_principals_to_OS_user_accounts[`hadoop.security.auth_to_local`].
```

File: docs/modules/hdfs/pages/usage-guide/security.adoc
Comment (adwk67) [inline docs/modules/hdfs/pages/usage-guide/security.adoc]:
```suggestion
Because of these issues we do not use a custom GroupMapper and only rely on the authorizer, which in turn receives a complete `UserGroupInformation` object, including the `shortUserName` and the precious full `userName`.
```

File: docs/modules/hdfs/pages/usage-guide/security.adoc
Comment (adwk67) [inline docs/modules/hdfs/pages/usage-guide/security.adoc]:
```suggestion
This has the downside that the group memberships used in OPA for authorization are not known to HDFS.
```

File: docs/modules/hdfs/pages/usage-guide/security.adoc
Comment (adwk67) [inline docs/modules/hdfs/pages/usage-guide/security.adoc]:
```suggestion
The implication is thus that you cannot add users to the `superuser` group, which is needed for certain administrative actions in HDFS.
```

File: docs/modules/hdfs/pages/usage-guide/security.adoc
Comment (adwk67) [inline docs/modules/hdfs/pages/usage-guide/security.adoc]:
```suggestion
We have decided that this is an acceptable approach as normal operations will not be affected.
```

**Review by adwk67 — CHANGES_REQUESTED** [review]
Just reviewed the docs so far with a few comments. Nit: we use regorule, rego rule and rego-rule here: I don't mind which it is but we should be consistent. The opa docs seem to use two separate words i.e. rego rules.

File: rust/operator-binary/src/security/kerberos.rs
Comment (adwk67) [inline rust/operator-binary/src/security/kerberos.rs:93]: I think we need to add a comment here to make this clear: are we assuming that any static mapping that **is** required, is the responsibility of the person writing the custom resource? Or is this not required anymore with the opa authorizer?

File: tests/templates/kuttl/kerberos/30-access-hdfs.yaml.j2
Comment (adwk67) [inline tests/templates/kuttl/kerberos/30-access-hdfs.yaml.j2]: should we comment out this line if it does not work?

File: tests/templates/kuttl/kerberos/12-rego-rules.yaml.j2
Comment (adwk67) [inline tests/templates/kuttl/kerberos/12-rego-rules.yaml.j2]: suggestion: if we call this via a file rather than via an inline kubectl apply it is easier to check indentation etc. - like it is done here: 
https://github.com/stackabletech/trino-operator/blob/main/tests/templates/kuttl/authentication/12-create-trino-cluster.yaml#L8

File: tests/templates/kuttl/kerberos/12-rego-rules.yaml.j2
Comment (adwk67) [inline tests/templates/kuttl/kerberos/12-rego-rules.yaml.j2]: maybe mention that the three `matches_identity` calls are OR-ed together?

File: rust/operator-binary/src/security/kerberos.rs
Comment (sbernauer) [reply rust/operator-binary/src/security/kerberos.rs:93]: I *think* it is not required any more, so I don't expect the user to do anything at all

File: tests/templates/kuttl/kerberos/30-access-hdfs.yaml.j2
Comment (sbernauer) [reply tests/templates/kuttl/kerberos/30-access-hdfs.yaml.j2]: Sure! 40884e4

File: tests/templates/kuttl/kerberos/12-rego-rules.yaml.j2
Comment (sbernauer) [reply tests/templates/kuttl/kerberos/12-rego-rules.yaml.j2]: substenv

File: tests/templates/kuttl/kerberos/12-rego-rules.yaml.j2
Comment (sbernauer) [reply tests/templates/kuttl/kerberos/12-rego-rules.yaml.j2]: Hmm I personally would argue that this is basic rego rule knowledge and would spoil the rules but not entirely sure what is better. The group in Karlsruhe decided to leave it as is and had an improvement in dba095f

File: rust/operator-binary/src/security/kerberos.rs
Comment (sbernauer) [reply rust/operator-binary/src/security/kerberos.rs:93]: Changed as discussed in bb264da

Comment (adwk67) [reply rust/operator-binary/src/security/kerberos.rs:93]: Thanks!

File: tests/templates/kuttl/kerberos/12-rego-rules.yaml.j2
Comment (sbernauer) [reply tests/templates/kuttl/kerberos/12-rego-rules.yaml.j2]: Implemented in efd0cec

**Comment by sbernauer** [comment]
@adwk67 feedback should be addressed

File: rust/operator-binary/src/security/opa.rs
Comment (NickLarsenNZ) [reply rust/operator-binary/src/security/opa.rs:1]: I'm happy with that, it tells me more than I knew about it before :)

**Comment by sbernauer** [comment]
Started https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/hdfs-operator-it-custom/129/

**Review by adwk67 — DISMISSED** [review]
LGTM! Can merge when the CI tests are all done.

**Review by adwk67 — APPROVED** [review]
LGTM! Can merge when the CI tests are all done.

**Comment by sbernauer** [comment]
And also a full testsuite run: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/hdfs-operator-it-custom/130/

**Comment by sbernauer** [comment]
Another full testsuite run, after I increased the resources in https://github.com/stackabletech/ci/commit/40937a9b0ac7c6b3271347aabb72c3493ba0aa8d:
https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/hdfs-operator-it-custom/131/

**Comment by sbernauer** [comment]
Full testsuite passed :rocket:

**Pull request merged**
